### PR TITLE
Account activation and password reset

### PR DESCRIPTION
**Pull request by flaviuvr**
Changes in this PR:

- Account activation and password reset functionalities using mailgun addon for heroku
- Tests for the cart and unactivated accounts
- Updated styling for every button

The app can be viewed deployed on heroku [here](https://canteenkoala.herokuapp.com/)

To access the cart functionality, you need to log in with the credentials:
 user: 'koala@food.org', password: 'easytoguesspassword' 

This caveat is due to mailgun being able to only send mails to authorized emails, thus the need for an already activated account.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -30,13 +33,46 @@ def session_token
...
-    return false if remember_digest.nil?
+  def authenticated?(attribute, token)
+    digest = send("#{attribute}_digest")
+    return false if digest.nil?
 
-    BCrypt::Password.new(remember_digest).is_password?(remember_token)
+    BCrypt::Password.new(digest).is_password?(token)
   end
 
   def forget
```
Comment (cristichirtos) [inline app/models/user.rb]: This method should have a bang, as it changes the internal state of the object.

File: app/models/user.rb [hunk 2]
```diff
@@ -30,13 +33,46 @@ def session_token
...
+
+  def activate!
+    update_columns(activated: true, activated_at: Time.zone.now)
+  end
+
+  def send_activation_email
+    UserMailer.account_activation(self).deliver_now
+  end
+
+  def create_reset_digest
```
Comment (cristichirtos) [inline app/models/user.rb]: Same for this method, although I didn't realize it either when I wrote this method, I'll modify mine too :laughing:

File: app/models/user.rb [hunk 3]
```diff
@@ -30,13 +33,46 @@ def session_token
...
+    reset_sent_at < 2.hours.ago
+  end
+
+  private
+
+  def downcase_email
+    email.downcase!
+  end
+
+  def create_activation_digest
```
Comment (cristichirtos) [inline app/models/user.rb]: And I think this one could use a bang too

File: spec/mailers/previews/user_mailer_preview.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
+# Preview all emails at http://localhost:3000/rails/mailers/user_mailer
+class UserMailerPreview < ActionMailer::Preview
+
+  # Preview this email at http://localhost:3000/rails/mailers/user_mailer/account_activation
+  def account_activation
+    user = User.last
```
Comment (cristichirtos) [inline spec/mailers/previews/user_mailer_preview.rb]: You can have `user = User.new` with a random email to not have to get data from the database for the previews and make it easier to know what it should display.

File: spec/mailers/previews/user_mailer_preview.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    user.activation_token = User.new_token
+    UserMailer.account_activation(user)
+  end
+
+  # Preview this email at http://localhost:3000/rails/mailers/user_mailer/password_reset
+  def password_reset
+    user = User.first
```
Comment (cristichirtos) [inline spec/mailers/previews/user_mailer_preview.rb]: Same here

File: config/routes.rb [hunk 6]
```diff
@@ -20,4 +22,7 @@
   resources :carts, only: :index do
     post :remove_cart_items, on: :collection
   end
+
+  resources :account_activations, only: [:edit]
```
Comment (cristichirtos) [inline config/routes.rb]: You don't need [ ] if there's only one action, and for the one below, you can use the `%i` like you did above.

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -18,11 +24,13 @@ def create
 
   def destroy
     log_out if logged_in?
+
     redirect_to root_url
   end
 
   def add_product(product)
```
Comment (cristichirtos) [inline app/controllers/sessions_controller.rb]: Why do you have an add_products method in the SessionsController class when in the routes this action is mapped to products#add_to_cart?

File: app/controllers/products_controller.rb [hunk 8]
```diff
@@ -48,6 +48,11 @@ def destroy
   end
 
   def add_to_cart
+    unless logged_in?
```
Comment (cristichirtos) [inline app/controllers/products_controller.rb]: You can extract this in a before_action, since it's an action that will likely be common to other controllers too, you can define it in ApplicationController, as the other controllers extend it.

**Review by cristichirtos — COMMENTED** [review]
Very good work, Flaviu! I especially love the website design, it looks very fresh. Nice PR description too!

File: config/routes.rb [hunk 9]
```diff
@@ -20,4 +22,7 @@
...
+  resources :password_resets, only: [:new, :create, :edit, :update]
```
Comment (sasaranAndrei) [inline config/routes.rb]: You can use `only: %i[new create edit update]` since there's a list of symbols.

File: app/controllers/sessions_controller.rb (hunk 7 above)
Comment (flaviuvr) [reply app/controllers/sessions_controller.rb]: Whoops, forgot about this😅

File: config/routes.rb [hunk 10]
```diff
@@ -1,4 +1,6 @@
 Rails.application.routes.draw do
+  get 'password_resets/new'
+  get 'password_resets/edit'
```
Comment (sasaranAndrei) [inline config/routes.rb]: Also, you can remove these lines since you used `resources :password_resets, only: [:new, :create, :edit, :update]`.

File: spec/mailers/previews/user_mailer_preview.rb [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+class UserMailerPreview < ActionMailer::Preview
+
+  # Preview this email at http://localhost:3000/rails/mailers/user_mailer/account_activation
+  def account_activation
+    user = User.last
+    user.activation_token = User.new_token
+    UserMailer.account_activation(user)
+  end
+
+  # Preview this email at http://localhost:3000/rails/mailers/user_mailer/password_reset
```
Comment (sasaranAndrei) [inline spec/mailers/previews/user_mailer_preview.rb:14]: I suggest you to remove these comments. It will make the code more readable.

Comment (flaviuvr) [reply spec/mailers/previews/user_mailer_preview.rb:14]: I prefer to keep these ones for the quick access they provide when testing these views

**Review by cristichirtos — APPROVED** [review]
Congrats, awesome work! :tada:

**Review by sasaranAndrei — APPROVED** [review]
Good job, Flaviu! The site looks very nice!

File: app/controllers/password_resets_controller.rb [hunk 12]
```diff
@@ -0,0 +1,61 @@
+class PasswordResetsController < ApplicationController
+  before_action :provide_user, only: [:edit, :update]
+  before_action :valid_user?, only: [:edit, :update]
```
Comment (vaneaclaudia) [inline app/controllers/password_resets_controller.rb]: by naming the method like this, it should return true/false but `valid_user?` method does not return true/false. It does either a redirect, either it returns nil

File: spec/requests/products_spec.rb [hunk 13]
```diff
@@ -0,0 +1,21 @@
+require 'rails_helper'
+
+RSpec.describe 'Products', type: :request do
```
Comment (vaneaclaudia) [inline spec/requests/products_spec.rb]: RSpec.describe should have the class name you are testing. Do you have a products.rb class?

So you should have it like:

`RSpec.describe PasswordResetsController, type: :request `

By having it like this, you will be able to call described_class helper method which will then be the instance of PasswordResetsController

File: spec/requests/products_spec.rb [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+
+  describe 'POST products#add_to_cart' do
+    let!(:product) { Product.create(title: 'Product', description: 'Details', price: 2) }
+
+    subject do
+      product.update_attribute(:id, 400)
```
Comment (vaneaclaudia) [inline spec/requests/products_spec.rb]: this should go in a before block

File: spec/requests/products_spec.rb [hunk 15]
```diff
@@ -0,0 +1,21 @@
+require 'rails_helper'
+
+RSpec.describe 'Products', type: :request do
+
+  describe 'POST products#add_to_cart' do
+    let!(:product) { Product.create(title: 'Product', description: 'Details', price: 2) }
```
Comment (vaneaclaudia) [inline spec/requests/products_spec.rb]: you could use `let` instead of `let!` since this will be called in subject

File: spec/requests/sessions_spec.rb [hunk 16]
```diff
@@ -0,0 +1,16 @@
+require 'rails_helper'
+
+RSpec.describe 'Sessions', type: :request do
```
Comment (vaneaclaudia) [inline spec/requests/sessions_spec.rb]: same remark regarding class name

File: app/controllers/password_resets_controller.rb [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+    elsif @user.update(user_params)
+      reset_session
+      @user.update_attribute(:reset_digest, nil)
+      log_in(@user)
+      flash[:success] = 'Password has been reset successfully!'
+
+      redirect_to @user
+    else
+      render 'edit'
+    end
```
Comment (vaneaclaudia) [inline app/controllers/password_resets_controller.rb:33]: could you maybe split up the code into different methods? It is hard to follow when there are multiple if elseif else branches

**Review by vaneaclaudia — CHANGES_REQUESTED** [review]

**Review by AdrianFlorian — APPROVED** [review]
👏 nice job!

File: app/controllers/account_activations_controller.rb [hunk 18]
```diff
@@ -0,0 +1,16 @@
+class AccountActivationsController < ApplicationController
+  def edit
+    user = User.find_by(email: params[:email])
+    if user && !user.activated? && user.authenticated?(:activation, params[:id])
```
Comment (VictorMotogna) [inline app/controllers/account_activations_controller.rb]: you could simplify this condition by adding safe operator `&.` to both conditions

File: app/controllers/password_resets_controller.rb [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+  def user_params
+    params.require(:user).permit(:password, :password_confirmation)
+  end
+
+  def provide_user
+    @user = User.find_by(email: params[:email])
+  end
+
+  def handle_valid_user
+    redirect_to root_path unless @user && @user.activated? && @user.authenticated?(:reset, params[:id])
```
Comment (VictorMotogna) [inline app/controllers/password_resets_controller.rb]: same here, condition could be simlpified

File: app/views/user_mailer/password_reset.text.haml [hunk 20]
```diff
@@ -0,0 +1,7 @@
+Hello,
+= @user.name
+To reset your password click the link below:
+= edit_password_reset_url(@user.reset_token, email: @user.email)
+
+The link will expire in 2 hours.
+If you did not request your password to be reset, please ignore this email.
```
Comment (VictorMotogna) [inline app/views/user_mailer/password_reset.text.haml]: you should add a newline at the end of this file

File: spec/factories.rb [hunk 21]
```diff
@@ -0,0 +1,15 @@
+FactoryBot.define do
+  factory :user do
+    name  { 'User test' }
```
Comment (VictorMotogna) [inline spec/factories.rb:3]: not required to change, but if you want you could take a look over [Faker](https://github.com/faker-ruby/faker) for mocking this data

File: spec/requests/products_spec.rb [hunk 22]
```diff
@@ -0,0 +1,17 @@
+require 'rails_helper'
+
+RSpec.describe ProductsController, type: :request do
+
+  describe 'POST products#add_to_cart' do
+    let(:product) { create :product }
+    subject { post add_to_cart_product_path(product.id) }
+
+    context 'not logged in' do
```
Comment (VictorMotogna) [inline spec/requests/products_spec.rb:12]: I think you could also add a test that succeeds, not only for unauthorized users

File: spec/requests/sessions_spec.rb [hunk 23]
```diff
@@ -0,0 +1,16 @@
+require 'rails_helper'
+
+RSpec.describe SessionsController, type: :request do
+
+  describe 'POST sessions#create' do
+    let(:user) { create :user}
+
+    context 'account not activated' do
```
Comment (VictorMotogna) [inline spec/requests/sessions_spec.rb:8]: same here, a test for a successful case, not only for not activated user

**Review by VictorMotogna — APPROVED** [review]
Good job!
